File: 2023/csharp/day03b/Program.cs
using System;
using System.Collections.Generic;
using System.Text;

class Program
{
    struct Position {
        public int Row;
        public int Col;

        public Position(int row, int col)
        {
            Row = row;
            Col = col;
        }
    }

    static void Usage()
    {
        string? progPath = Environment.ProcessPath;
        string? progName = Path.GetFileName(progPath);
        Console.WriteLine("usage: " + progName + " <file>");
        System.Environment.Exit(1);
    }

    static Dictionary<Position, string> BuildNumbers(string contents)
    {
        Dictionary<Position, string> numberLocs = new Dictionary<Position, string>();
        bool scanningNumber = false;
        StringBuilder number = new StringBuilder();
        Position currentPos = new Position();
        int row = 0;
        foreach ( string line in contents.Split('\n') )
        {
            for ( int col = 0; col < line.Length; col++ )
            {
                char ch = line[col];
                if ( scanningNumber ) {
                    if ( Char.IsDigit(ch) ) {
                        number.Append(ch);
                    } else {
                        numberLocs[currentPos] = number.ToString();
                        currentPos = new Position();
                        number = new StringBuilder();
                        scanningNumber = false;
                    }
                } else {
                    if ( Char.IsDigit(ch) ) {
                        number.Append(ch);
                        currentPos.Row = row;
                        currentPos.Col = col;
                        scanningNumber = true;
                    }
                }
            }
            if ( scanningNumber ) {
                numberLocs[currentPos] = number.ToString();
                currentPos = new Position();
                number = new StringBuilder();
                scanningNumber = false;
            }
            row++;
        }
        return numberLocs;
    }

    static Dictionary<Position, char> BuildGears(string contents)
    {
        Dictionary<Position, char> gearLocs = new Dictionary<Position, char>();
        int row = 0;
        foreach ( string line in contents.Split('\n') )
        {
            for ( int col = 0; col < line.Length; col++ )
            {
                char ch = line[col];
                if ( !Char.IsDigit(ch) && ch != '.' )
                {
                    Position currentPos = new Position(row, col);
                    gearLocs[currentPos] = ch;
                }
            }
            row++;
        }
        return gearLocs;
    }

    static uint CheckGears(Dictionary<Position, string> numberLocs, Dictionary<Position, char> gearLocs)
    {
        uint result = 0;
        foreach ( Position gearLoc in gearLocs.Keys )
        {
            List<uint> adjacents = new List<uint>();
            foreach ( Position numberLoc in numberLocs.Keys )
            {
                bool skip = false;
                for ( int deltaRow = -1; deltaRow <= 1; deltaRow++ )
                {
                    int adjacentRow = gearLoc.Row + deltaRow;
                    for ( int deltaCol = -1; deltaCol <= 1; deltaCol++ )
                    {
                        int adjacentCol = gearLoc.Col + deltaCol;
                        for ( int numberCol = numberLoc.Col; numberCol < numberLoc.Col + numberLocs[numberLoc].Length; numberCol++ )
                        {
                            if ( !skip && adjacentRow == numberLoc.Row && adjacentCol == numberCol )
                            {
                                skip = true;
                                adjacents.Add(UInt32.Parse(numberLocs[numberLoc]));
                            }
                        }
                    }
                }
            }
            if ( adjacents.Count == 2 )
            {
                uint prod = 1;
                foreach ( uint adjacent in adjacents )
                {
                    prod *= adjacent;
                }
                result += prod;
            }
        }
        return result;
    }

    static uint Process(string contents)
    {
        Dictionary<Position, string> numberLocs = BuildNumbers(contents);
        Dictionary<Position, char> gearLocs = BuildGears(contents);
        return CheckGears(numberLocs, gearLocs);
    }

    public static void Main(string[] args)
    {
        if ( args.Length < 1 )
        {
            Program.Usage();
        }
        string fileName = args[0];
        string contents = File.ReadAllText(fileName);
        contents = contents.Remove(contents.Length - 1, 1);
        uint result = Process(contents);
        Console.WriteLine("result = " + result);
    }
}

File: 2023/fsharp/day03b/Program.fs
open System
open System.IO

type Position = { Row : int; Col : int }

let usage () =
    let progname = Path.GetFileName(Environment.GetCommandLineArgs().[0])
    printfn "usage: %s <file>" progname
    Environment.Exit(1)

let buildNumbers (contents: string) =
    let processLine (row, numLocs) line =
        let processChar (col, scanningNumber, number, currentPos, numLocs) ch =
            if scanningNumber then
                if Char.IsDigit ch then
                    if col + 1 = String.length line then
                        col + 1, false, "", { Row = -1; Col = -1 }, numLocs |> Map.add currentPos (number + string ch)
                    else
                        col + 1, true, number + string ch, currentPos, numLocs
                else
                    col + 1, false, "", { Row = -1; Col = -1 }, numLocs |> Map.add currentPos number
            else
                if Char.IsDigit ch then
                    col + 1, true, string ch, { Row = row; Col = col }, numLocs
                else
                    col + 1, false, "", { Row = -1; Col = -1 }, numLocs
        let _, _, _, _, updatedNumLocs =
            Seq.fold processChar (0, false, "", { Row = -1; Col = -1 }, numLocs) line
        row + 1, updatedNumLocs
    let _, numLocs =
        Seq.fold processLine (0, Map.empty) (contents.Split('\n'))
    numLocs

let buildGears (contents: string) =
    let processLine (row, gearLocs) line =
        let processChar (col, gearLocs) ch =
            if ch = '*' then
                col + 1, gearLocs |> Map.add { Row = row; Col = col } ch
            else
                col + 1, gearLocs
        let _, updatedGearLocs =
            Seq.fold processChar (0, gearLocs) line
        row + 1, updatedGearLocs
    let _, gearLocs =
        Seq.fold processLine (0, Map.empty) (contents.Split('\n'))
    gearLocs

let checkGears (numberLocs : Map<Position, string>) (gearLocs : Map<Position, char>) =
    Seq.fold
        (fun result gearLoc ->
            let adjacents =
                Seq.fold
                    (fun adjacentsVec numberLoc ->
                        let numberRow = numberLoc.Row
                        let numberColFirst = numberLoc.Col
                        let numberColLast = numberLoc.Col + (String.length (Map.find numberLoc numberLocs))
                        let foundAdjacent =
                            Seq.fold
                                (fun found neighbor ->
                                    let adjacentPos = { Row = gearLoc.Row + neighbor.Row; Col = gearLoc.Col + neighbor.Col }
                                    if
                                        (adjacentPos.Row = numberRow)
                                        && (adjacentPos.Col >= numberColFirst)
                                        && (adjacentPos.Col < numberColLast)
                                    then
                                        true
                                    else
                                        found)
                                false
                                [
                                    { Row = -1; Col = -1 }; { Row = -1; Col = 0 }; { Row = -1; Col = 1 }
                                    { Row = 0; Col = -1 }; { Row = 0; Col = 1 }
                                    { Row = 1; Col = -1 }; { Row = 1; Col = 0 }; { Row = 1; Col = 1 }
                                ]

                        if foundAdjacent then
                            adjacentsVec @ [Map.find numberLoc numberLocs]
                        else
                            adjacentsVec)
                    []
                    (Map.keys numberLocs)

            if Seq.length adjacents = 2 then
                result + (Seq.fold (fun acc n -> acc * int n) 1 (Seq.map int adjacents))
            else
                result)
        0
        (Map.keys gearLocs)

let processInput (contents: string) =
    let numberLocs = buildNumbers(contents)
    let gearLocs = buildGears(contents)
    checkGears numberLocs gearLocs

[<EntryPoint>]
let main argv =
    if argv.Length < 1 then
        usage()
    let filename = argv.[0]
    let contents = File.ReadAllText(filename)
    let result = processInput contents
    printfn "result = %d" result
    0
